processor/k8sprocessor/processor.go

```diff
@@ -144,7 +176,63 @@ func (kp *kubernetesprocessor) ConsumeTraces(ctx context.Context, td pdata.Trace
 		}
 	}
 
-	return kp.nextConsumer.ConsumeTraces(ctx, td)
+	return kp.nextTraceConsumer.ConsumeTraces(ctx, td)
+}
+
+// ConsumeMetrics process metrics and add k8s metadata using resource hostname as pod origin.
+// TODO: Move to internal data model once it's available in contrib.
+func (kp *kubernetesprocessor) ConsumeMetrics(ctx context.Context, metrics pdata.Metrics) error {
+	mds := pdatautil.MetricsToMetricsData(metrics)
+
+	for i := range mds {
+		md := &mds[i]
+		var presetPodIP string
+		var podIP string
+
+		// Check if a collector/agent or a prior processor has already annotated the metrics with IP.
+		if md.Resource.GetLabels() != nil {
+			presetPodIP = md.Resource.Labels[k8sIPLabelName]
+		}
+
+		// Most of the metric receivers uses "host.hostname" resource label (which is represented as
+		// Node.Identifier.HostName in OpenCensus format) to identify metrics origin.
+		// In k8s environment, it's set to a pod IP address. If the value doesn't represent
+		// an IP address, we skip it.
+		if podIP == "" && md.Node.GetIdentifier().GetHostName() != "" {
+			hostname := md.Node.Identifier.HostName
+			if net.ParseIP(hostname) != nil {
+				podIP = hostname
+			}
+		}
+
+		if presetPodIP == "" && podIP != "" {
+			if md.Resource == nil {
+				md.Resource = &resourcepb.Resource{}
+			}
+			if md.Resource.Labels == nil {
+				md.Resource.Labels = map[string]string{}
+			}
+			md.Resource.Labels[k8sIPLabelName] = podIP
+		}
+
+		// Don't invoke any k8s client functionality in passthrough mode.
+		// Just tag the IP and forward the batch.
+		if kp.passthroughMode || podIP == "" {
+			continue
+		}
+
+		// Add k8s tags to resource.
+		attrsToAdd := kp.getAttributesForPodIP(podIP)
+		if len(attrsToAdd) == 0 {
+			continue
+		}
+
+		for k, v := range attrsToAdd {
+			md.Resource.Labels[k] = v
+		}
+	}
+
+	return kp.nextMetricsConsumer.ConsumeMetrics(ctx, metrics)
```

Comment:
The loop above modifies `mds` and here we pass `metrics`. It is not apparent that modifying `mds` changes `metrics` (does it?). I am not sure if anything can be done to make it clear. We probably just need to wait until internal metric data type is available and get rid of the conversion.

Reply:
yes, it does. it operates on pointers https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/358/files/141277673a43b49fba34c48cef1cae2a45780a43#diff-325890f0a975205128a62dd76a8392f9R188 to have the modification applied.